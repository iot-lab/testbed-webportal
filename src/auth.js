// import axios from 'axios'
import {iotlab} from './rest'

export const auth = {

  loggedIn: localStorage.getItem('loggedIn') || false,
  isAdmin: localStorage.getItem('isAdmin') || false,
  username: localStorage.getItem('username') || '',

  async doLogin (username, password) {
    iotlab.create(username, password)
    localStorage.setItem('apiAuth', JSON.stringify({
      username: username,
      password: password,
    }))

    await iotlab.getUserInfo(username)
    .then(user => {
      this.loggedIn = true
      this.username = username
      this.isAdmin = user.admin
      localStorage.setItem('loggedIn', this.loggedIn)
      localStorage.setItem('isAdmin', this.isAdmin)
      localStorage.setItem('username', this.username)
    })
    .catch(err => {
      console.log(err)
      this.isAdmin = false
      this.loggedIn = false
      this.username = ''
      throw err
    })
  },

  doLogout () {
    this.loggedIn = false
    localStorage.removeItem('loggedIn')
    localStorage.removeItem('apiAuth')
  },
}

