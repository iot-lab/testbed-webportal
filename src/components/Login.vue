
<template>
<div class="container mt-4">
  <div id="logo" class="mb-2 text-center mx-auto">
    <img src="../assets/disc-iotlab.svg">
  </div>
  <div class="login-box mx-auto text-center" :class="{ 'error': failed }">
    <h2 style="font-size: 1.75rem">Welcome to FIT IoT-LAB</h2>
    <form @submit.prevent="login" class="mx-auto">
      <input v-model="username" type="text" class="form-control" :class="{ 'is-invalid': failed }" placeholder="Username" autofocus required>
      <input v-model="password" type="password" class="form-control" :class="{ 'is-invalid': failed }" placeholder="Password" required>
      <div class="invalid-feedback mb-2">Invalid username or password</div>
      <button type="submit" class="btn btn-primary btn-block">Log in</button>
      <router-link to="signup" class="btn btn-outline-success btn-block">Register for an account</router-link>
      <p class="text-center">
        <router-link to="reset" class="btn btn-block">Forgot your password?</router-link>
      </p>
    </form>
  </div>
</div> <!-- container -->
</template>

<script>
import {auth} from '../auth'

export default {
  name: 'login',
  data () {
    return {
      username: '',
      password: '',
      failed: false,
    }
  },
  methods: {
    async login () {
      try {
        this.failed = false
        await auth.doLogin(this.username, this.password)
        this.$router.push('dashboard')
        // await userStore.login(this.email, this.password)
        // await sleep(200)
        // event.emit('user:loggedin')
      } catch (err) {
        this.failed = true
      }
    },
  },
}
</script>

<style scoped>

/**
 * Shake it!
 */
@keyframes shake {
  8%, 41% {
    transform: translateX(-10px);
  }
  25%, 58% {
    transform: translateX(10px);
  }
  75% {
    transform: translateX(-5px);
  }
  92% {
    transform: translateX(5px);
  }
  0%, 100% {
    transform: translateX(0);
  }
}

.login-box {
  padding: 15px;
  max-width: 450px;
}
.login-box.error {
  animation: shake .5s;
}

input {
  margin: 7px 0;
}

#logo img {
  width: 30%;
  max-width: 150px;
  opacity: 0.8;
}

form {
  max-width: 300px;
}
</style>
