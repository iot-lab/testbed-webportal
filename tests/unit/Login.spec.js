import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import Login from '@/views/Login'
import { auth } from '@/auth'
jest.mock('@/auth')

const localVue = createLocalVue()
localVue.use(VueRouter)
const router = new VueRouter()

describe('Login.vue', () => {

  it('should render blank login form', () => {
    const wrapper = shallowMount(Login, {localVue, router})
    expect(wrapper.element).toBeDefined()
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should redirect to dashboard after login', async () => {
    const wrapper = shallowMount(Login, {localVue, router})
    const login = 'login'
    const pwd = 'password'

    // fill form
    wrapper.find('input[type="text"]').setValue(login)
    wrapper.find('input[type="password"]').setValue(pwd)
    expect(wrapper.vm.username).toBe(login)
    expect(wrapper.vm.password).toBe(pwd)
    
    // submit form
    wrapper.find('form').trigger('submit')
    expect(auth.doLogin).toHaveBeenCalledWith(login, pwd)
    
    wrapper.vm.$router.push = jest.fn()
    wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.$router.push).toHaveBeenCalledWith('dashboard')
    })

  })

  it('should have failed state after bad login', async () => {
    const wrapper = shallowMount(Login, {localVue, router})

    auth.doLogin.mockImplementation(() => {
      throw new Error()
    })

    wrapper.setData({failed: false})
    wrapper.find('form').trigger('submit')
    expect(auth.doLogin).toHaveBeenCalledWith('', '')
    expect(wrapper.vm.failed).toBe(true)
    expect(wrapper.element).toBeDefined()
    expect(wrapper.element).toMatchSnapshot()
  })
  
})
