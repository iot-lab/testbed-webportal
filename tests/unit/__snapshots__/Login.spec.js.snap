// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login.vue should have failed state after bad login 1`] = `
<div
  class="container mt-4 text-center"
>
  <div
    class="mb-2 mx-auto"
    id="logo"
  >
    <img
      src="../assets/disc-iotlab.svg"
    />
  </div>
   
  <!---->
   
  <div
    class="login-box mx-auto error"
  >
    <h2
      style="font-size: 1.75rem;"
    >
      Welcome to FIT IoT-LAB
    </h2>
     
    <form
      class="mx-auto"
    >
      <input
        autofocus="autofocus"
        class="form-control is-invalid"
        placeholder="Username"
        required="required"
        type="text"
      />
       
      <input
        class="form-control is-invalid"
        placeholder="Password"
        required="required"
        type="password"
      />
       
      <div
        class="invalid-feedback mb-2"
      >
        Invalid username or password
      </div>
       
      <button
        class="btn btn-primary btn-block"
        type="submit"
      >
        Log in
      </button>
       
      <router-link-stub
        class="btn btn-outline-success btn-block"
        event="click"
        tag="a"
        to="[object Object]"
      >
        Register for an account
      </router-link-stub>
       
      <p
        class="text-center"
      >
        <router-link-stub
          class="btn btn-block"
          event="click"
          tag="a"
          to="[object Object]"
        >
          Forgot your password?
        </router-link-stub>
      </p>
    </form>
  </div>
</div>
`;

exports[`Login.vue should render blank login form 1`] = `
<div
  class="container mt-4 text-center"
>
  <div
    class="mb-2 mx-auto"
    id="logo"
  >
    <img
      src="../assets/disc-iotlab.svg"
    />
  </div>
   
  <!---->
   
  <div
    class="login-box mx-auto"
  >
    <h2
      style="font-size: 1.75rem;"
    >
      Welcome to FIT IoT-LAB
    </h2>
     
    <form
      class="mx-auto"
    >
      <input
        autofocus="autofocus"
        class="form-control"
        placeholder="Username"
        required="required"
        type="text"
      />
       
      <input
        class="form-control"
        placeholder="Password"
        required="required"
        type="password"
      />
       
      <div
        class="invalid-feedback mb-2"
      >
        Invalid username or password
      </div>
       
      <button
        class="btn btn-primary btn-block"
        type="submit"
      >
        Log in
      </button>
       
      <router-link-stub
        class="btn btn-outline-success btn-block"
        event="click"
        tag="a"
        to="[object Object]"
      >
        Register for an account
      </router-link-stub>
       
      <p
        class="text-center"
      >
        <router-link-stub
          class="btn btn-block"
          event="click"
          tag="a"
          to="[object Object]"
        >
          Forgot your password?
        </router-link-stub>
      </p>
    </form>
  </div>
</div>
`;
